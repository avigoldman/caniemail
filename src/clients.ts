import micromatch from 'micromatch';

import type { SupportType } from './json.cjs';

export const clientNames = [
  'apple-mail.macos',
  'apple-mail.ios',
  'gmail.desktop-webmail',
  'gmail.ios',
  'gmail.android',
  'gmail.mobile-webmail',
  'orange.desktop-webmail',
  'orange.ios',
  'orange.android',
  'outlook.windows',
  'outlook.windows-mail',
  'outlook.macos',
  'outlook.ios',
  'outlook.android',
  'yahoo.desktop-webmail',
  'yahoo.ios',
  'yahoo.android',
  'aol.desktop-webmail',
  'aol.ios',
  'aol.android',
  'samsung-email.android',
  'sfr.desktop-webmail',
  'sfr.ios',
  'sfr.android',
  'thunderbird.macos',
  'protonmail.desktop-webmail',
  'protonmail.ios',
  'protonmail.android',
  'hey.desktop-webmail',
  'mail-ru.desktop-webmail',
  'fastmail.desktop-webmail',
  'laposte.desktop-webmail'
] as const;

export type EmailClient = (typeof clientNames)[number];

type ClientSplit<S extends string> = S extends `${infer First}.${infer Second}`
  ? [First, Second]
  : never;
type ClientTuple = ClientSplit<(typeof clientNames)[number]>;
type Provider = ClientTuple[0];
type Platform = ClientTuple[1];

export type EmailClientGlobs = '*' | `${Provider}.${Platform}` | `*.${Platform}` | `${Provider}.*`;

export interface SupportTypeResult {
  noteNumbers: number[] | undefined;
  type: 'full' | 'partial' | 'none';
}

export const parseClients = (globs: EmailClientGlobs[]) =>
  Array.from(new Set<EmailClient>(micromatch(clientNames, globs) as any));

export function getSupportType(stats: Record<string, SupportType>): SupportTypeResult {
  const statKeys = Object.keys(stats).sort((k1, k2) => {
    if (k1 < k2) return -1;
    else if (k1 > k2) return 1;
    return 0;
  });

  // TODO: option to specify specific version, right now, it defaults to latest
  const latestEmailClient = statKeys[statKeys.length - 1];
  const supportStatus = stats[latestEmailClient];

  let noteNumbers: number[] | undefined = [...supportStatus.matchAll(/#(\d+)/g)]
    .map((matches) => matches[1])
    .map((noteNumber) => Number(noteNumber));

  if (noteNumbers.length === 0) noteNumbers = void 0;

  if (supportStatus.startsWith('y')) return { noteNumbers, type: 'full' };
  else if (supportStatus.startsWith('n')) return { noteNumbers, type: 'none' };

  return { noteNumbers, type: 'partial' };
}
