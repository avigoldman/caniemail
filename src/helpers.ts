import css, { type CssStylesheetAST } from '@adobe/css-tools';
import type { Document, Element, Node, Text } from 'domhandler';
import * as htmlparser from 'htmlparser2';

interface ParseHtmlResult {
  document: Document;
  stylesheets: CssStylesheetAST[];
}

export function fromTitleEntries<T>(
  entries: Array<{ title: string; value: T } | undefined>
): Record<string, T> {
  return Object.fromEntries(
    entries.filter((entry) => entry !== void 0).map((entry) => [entry.title, entry.value])
  );
}

export const getTitleMatches = (
  pairs: Record<string, any>,
  targetValue: string | string[],
  comparer?: (value: any, targetValue: string | string[]) => Boolean
): string[] =>
  Object.entries(pairs)
    .filter(([, value]) => (comparer ? comparer(value, targetValue) : targetValue.includes(value)))
    .map(([title]) => title);

const getStyleNodes = (node: Node) => {
  if (node.type === htmlparser.ElementType.Style) return [node as Element];

  const results: Element[] = [];

  if ('childNodes' in node) {
    for (const child of (node as Element).childNodes) {
      results.push(...getStyleNodes(child));
    }
  }

  return results;
};

export function parseHtml(html: string): ParseHtmlResult {
  const document = htmlparser.parseDocument(html);

  const styleNodes: Element[] = [];
  for (const childNode of document.childNodes) {
    styleNodes.push(...getStyleNodes(childNode));
  }

  const stylesheets: CssStylesheetAST[] = [];
  for (const styleNode of styleNodes) {
    const styleTextNode = styleNode.childNodes[0] as Text | undefined;
    if (styleTextNode !== void 0) {
      stylesheets.push(css.parse(styleTextNode.data));
    }
  }

  return {
    document,
    stylesheets
  };
}
