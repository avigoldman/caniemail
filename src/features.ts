import { getProperty } from 'dot-prop';
import onetime from 'onetime';

import json from './data/can-i-email.json';
import { getSupportType, type EmailClient } from './clients';

interface Feature {
  link: string;
  name: string;
  notes?: string[];
}

export interface FeatureIssue {
  notes: string[];
  title: string;
}

export interface FeatureIssues {
  errors: FeatureMap<FeatureIssue>;
  warnings: FeatureMap<FeatureIssue>;
}

type FeatureSet = Map<string, RawFeatureData>;

interface GetFeaturesResult {
  all: FeatureSet;
  css: FeatureSet;
  html: FeatureSet;
}

export type SupportType = string;

interface RawFeatureStats {
  aol: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
  };
  'apple-mail': {
    ios: Record<string, SupportType>;
    macos: Record<string, SupportType>;
  };
  fastmail: { 'desktop-webmail': Record<string, SupportType> };
  gmail: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
    'mobile-webmail': Record<string, SupportType>;
  };
  hey: { 'desktop-webmail': Record<string, SupportType> };
  laposte: { 'desktop-webmail': Record<string, SupportType> };
  'mail-ru': { 'desktop-webmail': Record<string, SupportType> };
  orange: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
  };
  outlook: {
    android: Record<string, SupportType>;
    ios: Record<string, SupportType>;
    macos: Record<string, SupportType>;
    'outlook-com': Record<string, SupportType>;
    windows: Record<string, SupportType>;
    'windows-mail': Record<string, SupportType>;
  };
  protonmail: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
  };
  'samsung-email': { android: Record<string, SupportType> };
  sfr: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
  };
  thunderbird: { macos: Record<string, SupportType> };
  yahoo: {
    android: Record<string, SupportType>;
    'desktop-webmail': Record<string, SupportType>;
    ios: Record<string, SupportType>;
  };
}

interface RawFeatureData {
  category: 'html' | 'css' | 'image';
  description: string | null;
  keywords: string | null;
  last_test_date: string;
  notes: string | null;
  notes_by_num: null | Record<string, string>;
  slug: string;
  stats: RawFeatureStats;
  test_results_url: string | null;
  test_url: string;
  title: string;
  url: string;
}

export interface RawData {
  api_version: string;
  data: RawFeatureData[];
  last_update_date: string;
  nice_names: any;
}

export const rawData = json as unknown as RawData;

const { data } = json as unknown as RawData;

export class FeatureMap<TValue> extends Map<EmailClient, TValue[]> {
  // @ts-ignore
  set(key: EmailClient, value: TValue): this {
    const values = super.get(key) || [];
    values.push(value);
    super.set(key, values);
    return this;
  }
}

export const getFeatures = onetime((): GetFeaturesResult => {
  const all = new Map();
  const css = new Map();
  const html = new Map();

  for (const bit of data) {
    all.set(bit.title, bit);
    if (bit.category === 'css') css.set(bit.title, bit);
    if (bit.category === 'html') html.set(bit.title, bit);
  }

  return {
    all,
    css,
    html
  };
});

export const getSupportedFeatures = (clients: EmailClient[]) => {
  const { all: features } = getFeatures();
  const supportedFeatures: Feature[] = [];

  for (const [featureTitle, featureData] of features) {
    const currentFeature: Feature = {
      link: featureData.url,
      name: featureTitle
    };
    let isFeatureSupported = true;

    for (const emailClient of clients) {
      const { stats } = featureData;
      const supportMap = getProperty(stats, emailClient);

      // eslint-disable-next-line no-continue
      if (supportMap === void 0) continue;

      const supportStatus = getSupportType(supportMap);

      if (supportStatus.type === 'none') {
        isFeatureSupported = false;
        break;
      }

      if (supportStatus.type === 'partial') {
        currentFeature.notes ??= [];

        for (const noteNumber of supportStatus.noteNumbers ?? []) {
          currentFeature.notes.push(
            `Note about \`${featureTitle}\` support for \`${emailClient}\`: ${
              featureData.notes_by_num![String(noteNumber)]
            }`
          );
        }
      }
    }

    if (isFeatureSupported) {
      supportedFeatures.push(currentFeature);
    }
  }

  return supportedFeatures;
};
