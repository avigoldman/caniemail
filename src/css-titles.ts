/* eslint-disable func-names */

import { getFeatures } from './features';
import { fromTitleEntries, getTitleMatches } from './helpers';

const features = getFeatures().css;
const keys = [...features.keys()];
const propertyNameRegex = /^[a-z-]+$/;

export const atRuleTitles = fromTitleEntries<string>(
  keys.map((title) => {
    if (!title.startsWith('@')) return void 0;

    return { title, value: title.replace(/^@/, '') };
  })
);

export const functionTitles = fromTitleEntries</* function name */ string>(
  keys.map((title) => {
    if (title.startsWith(':')) return void 0;

    if (title === 'CSS Variables (Custom Properties)') {
      return {
        title,
        value: 'var'
      };
    }

    if (!title.includes('()')) return void 0;

    const matches = /([a-z-]+)\(\)/.exec(title);
    if (matches === null) {
      throw new Error(`Could not determine function name in ${title}`);
    }

    return { title, value: matches[1] };
  })
);

export const keywordTitles = fromTitleEntries<string>(
  keys.map((title) => {
    if (!title.includes(' keyword')) return void 0;

    return {
      title,
      value: title.replace(/ keyword$/, '')
    };
  })
);

export const propertyValuePairTitles = fromTitleEntries<
  readonly [/* property */ string, /* value */ string]
>(
  keys.map((title) => {
    const matches = /([a-z-]+):\s*([a-z-]+)/.exec(title);
    if (matches === null) return void 0;

    return { title, value: [matches[1], matches[2]] as const };
  })
);

// Map of property titles to the properties they support
export const propertyTitles = fromTitleEntries<string[]>(
  keys.map((title) => {
    if (title === 'left, right, top, bottom') {
      return {
        title,
        value: ['left', 'right', 'top', 'bottom']
      };
    }

    const trimmedTitle = title
      .trim()
      .replace(/ shorthand$/, '')
      .replace(/ property$/, '');

    if (propertyNameRegex.test(trimmedTitle)) return { title, value: [trimmedTitle] };

    return void 0;
  })
);

export const unitTitles = fromTitleEntries<string>(
  keys.map((title) => {
    if (!title.endsWith(' unit')) return void 0;

    return { title, value: title.replace(/ unit$/, '') };
  })
);

export const getMatchingPropertyValuePairTitles = ({
  propertyName,
  propertyValue
}: {
  propertyName: string;
  propertyValue: string;
}): string[] =>
  getTitleMatches(
    propertyValuePairTitles,
    '',
    (value) => value[0] === propertyName && value[1] === propertyValue
  );

export const getMatchingAtRuleTitles = ({ atRules }: { atRules: string[] }) =>
  getTitleMatches(atRuleTitles, atRules);

export const getMatchingFunctionTitles = ({ propertyValue }: { propertyValue: string }) => {
  const matchingFunctionNames: string[] = [];

  // Match `<function>(` (e.g. `max(`, `calc(`)
  const valueFunctionNames = new Set(
    [...propertyValue.matchAll(/([a-z-]+)\(/g)].map((match) => match[1])
  );

  for (const [functionName, functionTitle] of Object.entries(functionTitles)) {
    if (valueFunctionNames.has(functionName)) {
      matchingFunctionNames.push(functionTitle);
    }

    valueFunctionNames.delete(functionName);
  }

  return matchingFunctionNames;
};

export const getMatchingKeywordTitles = ({ propertyValue }: { propertyValue: string }) =>
  getTitleMatches(keywordTitles, propertyValue);

export const getMatchingPropertyTitles = ({ propertyName }: { propertyName: string }): string[] =>
  getTitleMatches(propertyTitles, propertyName);

export const getMatchingUnitTitles = ({ propertyValue }: { propertyValue: string }) =>
  getTitleMatches(unitTitles, propertyValue, (value) => {
    if (value === 'initial') {
      return /\binitial\b/.test(propertyValue);
    }
    return new RegExp(`\\d${value}`).test(propertyValue);
  });
