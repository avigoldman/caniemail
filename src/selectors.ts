import * as CSSWhat from 'css-what';
import { SelectorType } from 'css-what';

import { getFeatures } from './features.js';
import { fromTitleEntries, getTitleMatches } from './helpers.js';

type SelectorDetector = (selectors: CSSWhat.Selector[][]) => boolean;

const features = getFeatures().css;
const keys = [...features.keys()];

export const selectorDetectorsMap: Record<string, SelectorDetector> = {
  'Adjacent sibling combinator': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Adjacent)) return true;
    }

    return false;
  },
  'Attribute selector': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (
        selector.some(
          (s) => s.type === SelectorType.Attribute && s.name !== 'id' && s.name !== 'class'
        )
      ) {
        return true;
      }
    }

    return false;
  },
  'Chaining selectors': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (
        selector.filter((s) => s.type === SelectorType.Attribute && s.name === 'class').length >= 2
      ) {
        return true;
      }
    }

    return false;
  },
  'Child combinator': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Child)) return true;
    }

    return false;
  },
  'Class selector': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Attribute && s.name === 'class')) {
        return true;
      }
    }

    return false;
  },
  'Descendant combinator': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Descendant)) return true;
    }

    return false;
  },
  'General sibling combinator': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Sibling)) {
        return true;
      }
    }

    return false;
  },
  'Grouping selectors': (selectors: CSSWhat.Selector[][]) => selectors.length >= 2,
  'ID selector': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Attribute && s.name === 'id')) {
        return true;
      }
    }

    return false;
  },
  'Type selector': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Tag)) return true;
    }

    return false;
  },
  'Universal selector *': (selectors: CSSWhat.Selector[][]) => {
    for (const selector of selectors) {
      if (selector.some((s) => s.type === SelectorType.Universal)) {
        return true;
      }
    }

    return false;
  }
};

export const psuedoSelectorTitles = fromTitleEntries<string>(
  keys.map((title) => {
    if (!title.startsWith(':')) return void 0;

    return { title, value: title };
  })
);

export const selectorTitles = fromTitleEntries<SelectorDetector>(
  keys.map((title) => {
    if (selectorDetectorsMap[title] === void 0) return void 0;

    return { title, value: selectorDetectorsMap[title] };
  })
);

export const getMatchingSelectorTitles = ({ selector }: { selector: string }) => {
  const parsedSelectors = CSSWhat.parse(selector);
  return getTitleMatches(selectorTitles, selector, (tester) => tester(parsedSelectors));
};

export const getMatchingPseudoSelectorTitles = ({ selector }: { selector: string }) =>
  getTitleMatches(psuedoSelectorTitles, selector);
