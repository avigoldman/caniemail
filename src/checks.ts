import type { CssStylesheetAST } from '@adobe/css-tools';
import type { Document, Element } from 'domhandler';
import { getProperty } from 'dot-prop';
import { ElementType } from 'htmlparser2';
import styleToObject from 'style-to-object';

import { getSupportType, type EmailClient } from './clients.js';
import {
  getMatchingAtRuleTitles,
  getMatchingFunctionTitles,
  getMatchingKeywordTitles,
  getMatchingPropertyTitles,
  getMatchingPropertyValuePairTitles,
  getMatchingUnitTitles
} from './css-titles.js';
import { getFeatures, type FeatureIssues } from './features.js';
import {
  getMatchingAttributeTitles,
  getMatchingElementAttributePairTitles,
  getMatchingElementTitles
} from './html-titles.js';
import { getMatchingPseudoSelectorTitles, getMatchingSelectorTitles } from './selectors.js';

interface BaseCheckArgs {
  clients: EmailClient[];
  issues: FeatureIssues;
}

interface CheckDeclarationsArgs extends BaseCheckArgs {
  declarations: Array<{ property: string; value: string }>;
}

interface CheckHtmlArgs extends BaseCheckArgs {
  document: Document;
}

interface CheckHtmlNodeArgs extends BaseCheckArgs {
  node: Element;
}

interface CheckFeaturesArgs extends BaseCheckArgs {
  titles: string | string[];
}

interface CheckSelectorsArgs extends BaseCheckArgs {
  selectors: string[];
}

interface CheckStylesheetArgs extends BaseCheckArgs {
  stylesheet: CssStylesheetAST;
}

const atRules = new Set([
  'charset',
  'custom-media',
  'document',
  'font-face',
  'host',
  'import',
  'keyframes',
  'keyframe',
  'media',
  'namespace',
  'page',
  'supports'
]);

const checkFeatures = ({ clients, issues, titles }: CheckFeaturesArgs) => {
  const { all: features } = getFeatures();

  for (const title of [titles].flat()) {
    for (const client of clients) {
      const feature = features.get(title);

      if (feature === void 0) throw new RangeError(`Feature \`${title}\` not found.`);

      const { stats } = feature;
      const supportMap = getProperty(stats, client);

      if (supportMap === void 0) {
        throw new RangeError(`Feature \`${title}\` not found on \`${client}\`.`);
      }

      const supportStatus = getSupportType(supportMap);
      const notes = (supportStatus.noteNumbers ?? []).map(
        (noteNumber) => feature.notes_by_num![String(noteNumber)]
      );

      if (supportStatus.type === 'none') {
        issues.errors.set(client, { notes, title });
      } else if (supportStatus.type === 'partial') {
        issues.warnings.set(client, { notes, title });
      }
    }
  }
};

const checkDeclarations = ({ clients, declarations, issues }: CheckDeclarationsArgs) => {
  const { css: features } = getFeatures();

  for (const declaration of declarations) {
    const { property: propertyName, value: propertyValue } = declaration;

    // Check that the property name is supported
    if (propertyName !== void 0 && features.get(propertyName) !== void 0) {
      const propertyTitles = getMatchingPropertyTitles({ propertyName });
      checkFeatures({ clients, issues, titles: propertyTitles });
    }

    // Check that the units and functions in the property value are supported
    if (propertyValue !== void 0) {
      const functionTitles = getMatchingFunctionTitles({ propertyValue });
      const keywordTitles = getMatchingKeywordTitles({ propertyValue });
      const unitTitles = getMatchingUnitTitles({ propertyValue });

      checkFeatures({ clients, issues, titles: functionTitles });
      checkFeatures({ clients, issues, titles: keywordTitles });
      checkFeatures({ clients, issues, titles: unitTitles });
    }

    // Check that the property name + value pair is supported
    if (propertyName !== void 0 && propertyValue !== void 0) {
      const propertyValuePairTitles = getMatchingPropertyValuePairTitles({
        propertyName,
        propertyValue
      });
      checkFeatures({ clients, issues, titles: propertyValuePairTitles });
    }
  }
};

const checkSelectors = ({ clients, issues, selectors }: CheckSelectorsArgs) => {
  for (const selector of selectors) {
    const pseudoSelectorTitles = getMatchingPseudoSelectorTitles({ selector });
    const mselectorTitles = getMatchingSelectorTitles({ selector });

    checkFeatures({ clients, issues, titles: pseudoSelectorTitles });
    checkFeatures({ clients, issues, titles: mselectorTitles });
  }
};

export const checkStylesheet = ({ clients, issues, stylesheet }: CheckStylesheetArgs) => {
  const matchedAtRules: string[] = [];
  for (const stylesheetRule of stylesheet.stylesheet?.rules ?? []) {
    if (stylesheetRule.type === 'rule') {
      const rule = stylesheetRule;
      const declarations = (rule.declarations ?? [])
        .filter((declaration) => declaration.type !== 'comment')
        .map((declaration) => {
          return {
            property: declaration.property,
            value: declaration.value
          };
        });

      checkDeclarations({ clients, declarations, issues });
      checkSelectors({ clients, issues, selectors: rule.selectors ?? [] });
    }

    if (atRules.has(stylesheetRule.type)) {
      matchedAtRules.push(stylesheetRule.type);
    }
  }

  const atRuleTitles = getMatchingAtRuleTitles({ atRules: matchedAtRules });

  checkFeatures({ clients, issues, titles: atRuleTitles });
};

const checkHtmlNode = ({ clients, issues, node }: CheckHtmlNodeArgs) => {
  const elementTitles = getMatchingElementTitles({ tagName: node.tagName });

  checkFeatures({ clients, issues, titles: elementTitles });

  if (node.attributes !== void 0) {
    const elementAttributePairTitles = getMatchingElementAttributePairTitles({
      attributes: Object.fromEntries(node.attributes.map((attr) => [attr.name, attr.value])),
      tagName: node.tagName
    });
    const attributeTitles = getMatchingAttributeTitles({
      attributes: node.attributes.map((attr) => attr.name)
    });

    checkFeatures({ clients, issues, titles: attributeTitles });
    checkFeatures({ clients, issues, titles: elementAttributePairTitles });

    // Check inline styles
    const styleAttr = node.attributes.find((attr) => attr.name === 'style');
    if (styleAttr !== void 0) {
      const styleObject = ((styleToObject as any).default ?? styleToObject)(styleAttr.value);
      if (styleObject !== null) {
        const declarations = Object.entries(styleObject).map(
          ([property, value]) => ({ property, value }) as any
        );
        checkDeclarations({ clients, declarations, issues });
      }
    }
  }

  if ('childNodes' in node) {
    for (const childNode of node.childNodes) {
      if (childNode.type === ElementType.Tag) {
        checkHtmlNode({ clients, issues, node: childNode as Element });
      }
    }
  }
};

export const checkHtml = ({ clients, issues, document }: CheckHtmlArgs) => {
  for (const childNode of document.childNodes) {
    if (childNode.type === ElementType.Tag) {
      checkHtmlNode({ clients, issues, node: childNode as Element });
    }
  }
};
