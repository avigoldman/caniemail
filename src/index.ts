import micromatch from 'micromatch';

import { checkHtml, checkStylesheet } from './checks';
import { clientNames, type EmailClient, type EmailClientGlobs } from './clients';
import { FeatureMap, type FeatureIssues, type FeatureIssue } from './features';
import { parseHtml } from './helpers';

export { rawData } from './features';

export interface CanIEmailOptions {
  /**
		An array of client names or globs to match email clients.
		Example: ['gmail.android', 'outlook.*', '*.ios']
	*/
  clients: EmailClientGlobs[];
  code: string;
}

export interface CanIEmailResult {
  issues: FeatureIssues;
  success: boolean;
}

export const caniemail = ({ clients: globs, code }: CanIEmailOptions): CanIEmailResult => {
  const { document, stylesheets } = parseHtml(code);
  const clients = Array.from(new Set<EmailClient>(micromatch(clientNames, globs) as any));
  const issues: FeatureIssues = {
    errors: new FeatureMap<FeatureIssue>(),
    warnings: new FeatureMap<FeatureIssue>()
  };

  if (clients.length === 0) {
    throw new RangeError(`The specified email client(s) (${globs.join(', ')}) were not found`);
  }

  for (const stylesheet of stylesheets) checkStylesheet({ clients, issues, stylesheet });

  checkHtml({ clients, document, issues });

  return {
    issues,
    success: issues.errors.size === 0
  };
};

interface FormatIssueOptions {
  client: EmailClient;
  issue: FeatureIssue;
  issueType: 'error' | 'warning';
}

export const formatIssue = ({ client, issue, issueType }: FormatIssueOptions) => {
  const { notes, title } = issue;
  return {
    message:
      issueType === 'error'
        ? `\`${title}\` is not supported by \`${client}\``
        : `\`${title}\` is only partially supported by \`${client}\``,
    notes: notes.map((note) => `Note about \`${title}\` support for \`${client}\`: ${note}`)
  };
};
