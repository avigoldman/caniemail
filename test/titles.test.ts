import { describe, expect, test } from 'vitest';

import {
  atRuleTitles,
  functionTitles,
  keywordTitles,
  propertyTitles,
  propertyValuePairTitles,
  unitTitles
} from '../src/css-titles';
import { attributeTitles, elementAttributePairTitles, elementTitles } from '../src/html-titles';
import { psuedoSelectorTitles, selectorTitles } from '../src/selectors';

describe('css', () => {
  Object.entries({
    atRuleTitles,
    functionTitles,
    keywordTitles,
    propertyTitles,
    propertyValuePairTitles,
    psuedoSelectorTitles,
    selectorTitles,
    unitTitles
  }).forEach(([name, map]) => {
    test(name, () => {
      expect(Object.keys(map).sort()).toMatchSnapshot();
    });
  });
});

describe('html', () => {
  Object.entries({
    attributeTitles,
    elementAttributePairTitles,
    elementTitles
  }).forEach(([name, map]) => {
    test(name, () => {
      expect(Object.keys(map).sort()).toMatchSnapshot();
    });
  });
});
