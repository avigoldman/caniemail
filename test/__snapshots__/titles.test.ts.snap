// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css > atRuleTitles 1`] = `
[
  "@font-face",
  "@import",
  "@keyframes",
  "@media",
  "@media (-webkit-device-pixel-ratio)",
  "@media (hover), @media (any-hover)",
  "@media (orientation)",
  "@media (prefers-color-scheme)",
  "@media (prefers-reduced-motion)",
  "@supports",
]
`;

exports[`css > functionTitles 1`] = `
[
  "CSS Variables (Custom Properties)",
  "CSS calc() function",
  "clamp()",
  "conic-gradient()",
  "lang()",
  "lch(), oklch(), lab(), oklab()",
  "light-dark()",
  "linear-gradient()",
  "max()",
  "min()",
  "radial-gradient()",
  "rgb()",
  "rgba()",
]
`;

exports[`css > keywordTitles 1`] = `
[
  "!important keyword",
]
`;

exports[`css > propertyTitles 1`] = `
[
  "accent-color",
  "align-items",
  "animation",
  "aspect-ratio",
  "backdrop-filter",
  "background",
  "background-blend-mode",
  "background-clip",
  "background-color",
  "background-image",
  "background-origin",
  "background-position",
  "background-repeat",
  "background-size",
  "border",
  "border-collapse",
  "border-image",
  "border-radius",
  "border-spacing",
  "box-shadow",
  "box-sizing",
  "caption-side",
  "clip-path",
  "column-count",
  "direction",
  "display",
  "empty-cells",
  "filter",
  "float",
  "font shorthand",
  "font-kerning",
  "font-size",
  "font-stretch",
  "font-weight",
  "height property",
  "hyphenate-character",
  "hyphenate-limit-chars",
  "hyphens",
  "inline-size ",
  "inset",
  "justify-content",
  "left, right, top, bottom",
  "letter-spacing",
  "line-height",
  "list-style",
  "list-style-image",
  "list-style-position",
  "list-style-type",
  "margin",
  "max-block-size",
  "max-height property",
  "max-inline-size",
  "max-width",
  "min-block-size",
  "min-height property",
  "min-inline-size",
  "min-width property",
  "mix-blend-mode",
  "object-fit",
  "object-position",
  "opacity",
  "orphans",
  "outline",
  "outline-offset",
  "overflow",
  "overflow-wrap",
  "padding",
  "position",
  "resize",
  "scroll-snap",
  "shape-margin",
  "shape-outside",
  "tab-size",
  "table-layout",
  "text-align",
  "text-align-last",
  "text-decoration",
  "text-decoration-color",
  "text-decoration-line",
  "text-decoration-skip-ink",
  "text-decoration-style",
  "text-decoration-thickness",
  "text-emphasis",
  "text-emphasis-position",
  "text-indent",
  "text-justify",
  "text-orientation",
  "text-overflow",
  "text-shadow",
  "text-transform",
  "text-underline-offset",
  "text-underline-position",
  "text-wrap",
  "transform",
  "transition",
  "user-select",
  "vertical-align",
  "visibility",
  "white-space",
  "white-space-collapse",
  "widows",
  "width property",
  "word-break",
  "word-spacing",
  "writing-mode",
  "z-index",
]
`;

exports[`css > propertyValuePairTitles 1`] = `
[
  "display:flex",
  "display:grid",
  "display:none",
  "flex-direction:column",
  "flex-wrap: wrap",
]
`;

exports[`css > psuedoSelectorTitles 1`] = `
[
  "::after",
  "::before",
  "::first-letter",
  "::first-line",
  "::marker",
  "::placeholder",
  ":active",
  ":checked",
  ":first-child",
  ":first-of-type",
  ":focus",
  ":has()",
  ":hover",
  ":last-child",
  ":last-of-type",
  ":link",
  ":not",
  ":nth-child",
  ":nth-last-child",
  ":nth-last-of-type",
  ":nth-of-type",
  ":only-child",
  ":only-of-type",
  ":target",
  ":visited",
]
`;

exports[`css > selectorTitles 1`] = `
[
  "Adjacent sibling combinator",
  "Attribute selector",
  "Chaining selectors",
  "Child combinator",
  "Class selector",
  "Descendant combinator",
  "General sibling combinator",
  "Grouping selectors",
  "ID selector",
  "Type selector",
  "Universal selector *",
]
`;

exports[`css > unitTitles 1`] = `
[
  "% unit",
  "ch unit",
  "cm unit",
  "em unit",
  "ex unit",
  "in unit",
  "initial unit",
  "mm unit",
  "pc unit",
  "pt unit",
  "px unit",
  "rem unit",
  "vh unit",
  "vmax unit",
  "vmin unit",
  "vw unit",
]
`;

exports[`html > attributeTitles 1`] = `
[
  "align attribute",
  "aria-describedby attribute",
  "aria-hidden attribute",
  "aria-label attribute",
  "aria-labelledby attribute",
  "aria-live attribute",
  "background attribute",
  "cellpadding attribute",
  "cellspacing attribute",
  "dir attribute",
  "height attribute",
  "hidden attribute",
  "lang attribute",
  "loading attribute",
  "popover attribute",
  "required attribute",
  "role attribute",
  "srcset and sizes attributes",
  "target attribute",
  "valign attribute",
  "width attribute",
]
`;

exports[`html > elementAttributePairTitles 1`] = `
[
  "<button type="reset"> element",
  "<button type="submit"> element",
  "<input type="checkbox"> element",
  "<input type="hidden"> element",
  "<input type="radio"> element",
  "<input type="reset"> element",
  "<input type="submit"> element",
  "<input type="text"> element",
  "Local anchors",
  "mailto: links",
]
`;

exports[`html > elementTitles 1`] = `
[
  "<abbr> element",
  "<acronym> element",
  "<audio> element",
  "<base>",
  "<bdi> element",
  "<blockquote> element",
  "<body> element",
  "<code> element",
  "<del> element",
  "<dfn> element",
  "<dialog> element",
  "<div> element",
  "<form> element",
  "<h1> to <h6> elements",
  "<hr> element",
  "<img> element",
  "<link> element",
  "<marquee> element",
  "<meter> element",
  "<object> element",
  "<p> element",
  "<picture> element",
  "<pre> element",
  "<progress> element",
  "<rp> element",
  "<rt> element",
  "<ruby> element",
  "<select> element",
  "<small> element",
  "<span> element",
  "<strike> element",
  "<strong> element",
  "<style> element",
  "<table> element",
  "<textarea> element",
  "<ul>, <ol> and <dl>",
  "<video> element",
  "<wbr> element",
  "Embedded <svg> image",
  "HTML5 doctype",
  "HTML5 semantics",
  "Image maps",
  "address",
]
`;
